cache/network/api.go

```diff
@@ -140,24 +155,156 @@ func (c apiClient) uploadArchive(archivePath, uploadMethod, uploadURL string, he
 	if err != nil {
 		c.logger.Warnf("error while dumping response: %s", err)
 	}
-	c.logger.Debugf("Response dump: %s", string(dump))
+	c.logger.Debugf("Chunk response dump: %s", string(dump))
 
 	if resp.StatusCode != http.StatusOK {
-		return unwrapError(resp)
+		return "", unwrapError(resp)
+	}
+
+	etag := resp.Header.Get("ETag")
+	fmt.Printf("Etag %s for %s of size %v \n\r", etag, uploadURL.URL, size)
+
+	return etag, nil
+}
+
+func (c apiClient) uploadArchive(archivePath string, chunkSize, chunkCount, lastChunkSize int, uploadURLs []uploadURL) ([]string, error) {
+
+	fmt.Printf("Uploading archive: %s \n\r", archivePath)
+	fmt.Printf("Chunk size, count, last chunk size: %d, %d, %d \n\r", chunkSize, chunkCount, lastChunkSize)
+	fmt.Printf("Upload urls: %+v \n\r", uploadURLs)
+
+	file, err := os.Open(archivePath)
+	if err != nil {
+		return nil, fmt.Errorf("open file: %s", err)
+	}
+	defer file.Close()
+
+	if chunkCount == 1 {
+		fileInfo, err := os.Stat(archivePath)
+		if err != nil {
+			return nil, fmt.Errorf("stat file: %s", err)
+		}
+		c.logger.Debugf("Uploading single chunk (non multipart upload)", uploadURLs[0].URL)
+
+		_, err = c.uploadArchiveChunk(uploadURLs[0], file, fileInfo.Size())
+		if err != nil {
+			return nil, fmt.Errorf("upload single chunk (non multipart): %s", err)
+		}
+		return nil, nil
+	}
+
+	type Job struct {
+		ChunkNumber int
+		ChunkStart  int64
+		ChunkSize   int64
+		UploadURL   uploadURL
+	}
+
+	jobs := make(chan Job, chunkCount)
+	results := make(chan struct {
+		ChunkNumber int
+		ChunkSize   int64
+		ETag        string
+	}, chunkCount)
+	errors := make(chan error, chunkCount)
```

Comment:
You don't need a chunkCount capacity channel for this, just have one channel and at the end try to select from it to chek if there's been any errors.

Also, cancel the context at the first error so that other workers stop downloading other chunks if their work will be discarded anyways